src/journal/compress.c

```diff
@@ -231,15 +257,74 @@ int decompress_blob_lz4(const void *src, uint64_t src_size,
 #endif
 }
 
-int decompress_blob(int compression,
-                    const void *src, uint64_t src_size,
-                    void **dst, size_t *dst_alloc_size, size_t* dst_size, size_t dst_max) {
+int decompress_blob_zstd(
+                const void *src, uint64_t src_size,
+                void **dst, size_t *dst_alloc_size, size_t* dst_size, size_t dst_max) {
+
+#if HAVE_ZSTD
+        size_t space;
+
+        assert(src);
+        assert(src_size > 0);
+        assert(dst);
+        assert(dst_alloc_size);
+        assert(dst_size);
+        assert(*dst_alloc_size == 0 || *dst);
+
+        if (src_size > SIZE_MAX/2) /* Overflow? */
+                return -ENOBUFS;
+        space = src_size * 2;
+        if (dst_max > 0 && space > dst_max)
+                space = dst_max;
+
+        if (!greedy_realloc(dst, dst_alloc_size, space, 1))
```

Comment:
not sure i grok the question, but the buffer passed might already be allocated to some size and we'll increase it to the size we need if necessary, and reuse it otherwise? hence, if the uncompressed size is known we'll enlarge the buffer as needed, and if it isn't we ensure it's double the size of what the input is, which is most likely going to work.

Reply:
> if the uncompressed size is known we'll enlarge the buffer as needed

Fine so far.

>  and if it isn't we ensure it's double the size of what the input is

...and decompress repeatedly from the start until the dest buffer fits. Id expect the streaming API being a better fit here.

Reply:
well, this is about blob decompression, really, i.e. stuff that fits fully into memory, and not arbitrary sized stuff.

We have decompress_stream() as generic wrapper for stream decompression, and that's what should be used for arbitrary sized stuff. But for "blob" decompression it shouldn#t really matter much. I mean sure, if you want to optimize a corner case, knock yourself out and send a follow-up patch, but most likely if blob compression is used for large stuff where this corner case is hit, then it's indication that blob compression should never have been used but stream compression instead, i.e. decompress_stream_xyz() instead of decompress_blob_xyz()...